internal/controller/nonadmindownloadrequest_controller.go

```diff
@@ -0,0 +1,292 @@
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the NonAdminDownloadRequest object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.20.2/pkg/reconcile
+//
+// This reconcile implements ObjectReconciler https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.20.2/pkg/reconcile#ObjectReconciler
+// Each reconciliation event gets the associated object from Kubernetes before passing it to Reconcile
+func (r *NonAdminDownloadRequestReconciler) Reconcile(ctx context.Context, req *nacv1alpha1.NonAdminDownloadRequest) (reconcile.Result, error) {
+	if req == nil || req.Spec.Target.Kind == constant.EmptyString {
+		return ctrl.Result{Requeue: false}, nil
+	}
+	// TODO: delete if UID is always available here.
+	// if req.ObjectMeta.UID  == "" {
+	// requeue later until uid is populated
+	// }
+	logger := log.FromContext(ctx)
+	logger.Info("Reconciling NonAdminDownloadRequest", "name", req.Name, "namespace", req.Namespace)
+	// defines associated downloadrequest for getting, or deleting
+	veleroDR := velerov1.DownloadRequest{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      req.VeleroDownloadRequestName(),
+			Namespace: r.OADPNamespace,
+			Labels: func() map[string]string {
+				nal := function.GetNonAdminLabels()
+				nal[constant.NadrOriginNACUUIDLabel] = string(req.GetUID())
+				return nal
+			}(),
+			Annotations: function.GetNonAdminDownloadRequestAnnotations(req),
+		},
+		Spec: velerov1.DownloadRequestSpec{
+			Target: velerov1.DownloadTarget{
+				Kind: req.Spec.Target.Kind,
+			},
+		},
+	}
+	// if request has status, it may already be processed or expired
+	if req.Status.VeleroDownloadRequest.Status != nil {
+		// if request is expired, delete NADR after deleting velero DR
+		if req.Status.VeleroDownloadRequest.Status.Expiration != nil && req.Status.VeleroDownloadRequest.Status.Expiration.Before(&metav1.Time{Time: time.Now()}) {
+			// find associated velero downloadrequest and delete that first
+			logger.V(1).Info("Deleting expired NonAdminDownloadRequest associated velero download request", req.VeleroDownloadRequestName(), req.Namespace)
+			if err := r.Delete(ctx, &veleroDR); err != nil {
+				if !apierrors.IsNotFound(err) {
+					logger.V(1).Info("Failed to delete expired NonAdminDownloadRequest associated velero download request", req.VeleroDownloadRequestName(), err)
+					// other errors, requeue to retry delete
+					return reconcile.Result{}, err
+				}
+			}
+			logger.V(1).Info("Deleting expired NonAdminDownloadRequest", req.Name, req.Namespace)
+			if err := r.Delete(ctx, req); err != nil {
+				if !apierrors.IsNotFound(err) {
+					logger.V(1).Info("Failed to delete expired NonAdminDownloadRequest", req.Name, err)
+					// other errors, requeue to retry delete
```

Comment:
What happes when there are error and the object will actually get removed by next reconcile?

Reply:
Yes.

Reply:
Or they will see in debug logs the "other error"

Reply:
they = admin, users won't notice?

Reply:
could add to condition. is that what you want?